src/idl_parser.cpp

```diff
@@ -845,6 +845,23 @@ CheckedError Parser::ParseField(StructDef &struct_def) {
   field->required = field->attributes.Lookup("required") != nullptr;
   if (field->required && (struct_def.fixed || IsScalar(type.base_type)))
     return Error("only non-scalar fields in tables may be 'required'");
+
+  // Mark the optional scalars. Note that a side effect of ParseSingleValue is
+  // fixing field->value.constant to null.
+  if (IsScalar(type.base_type)) {
+    field->optional = (field->value.constant == "null");
+    if (field->optional && !SupportsOptionalScalars()) {
+      return Error(
+        "Optional scalars are not yet supported in at least one the of "
+        "the specified programming languages."
+      );
+    }
+  } else {
+    // For nonscalars, only required fields are non-optional.
+    // At least until https://github.com/google/flatbuffers/issues/6053
+    field->optional = !field->required;
+  }
+
```

Comment:
@CasperN This is what is causing the issue (since those accessors aren't supposed to be optional) @aardappel correct me if I am wrong
```swift
/// these two are supposed to always return a value? since arrays don't have optional types
/// `inventory: [UInt8] = false;` will always cause an error
var inventoryCount: Int32 {}
var getInventory(at: Int32) -> UInt8 {}
```

Reply:
Yes, this is a language/API issue, I guess we return `0` for the length in Java, and that has been adopted by other languages where we can't return a cheap accessor. That is technically wrong though.